modules/bmc/redfish.rb

```diff
@@ -0,0 +1,210 @@
+      end
+
+      def poweroff(soft=false)
+        poweraction( soft ? 'GracefulShutdown' : 'ForceOff' )
+      end
+
+      def powercycle
+        poweraction('ForceRestart')
+      end
+
+      def poweron
+        poweraction('On')
+      end
+
+      def powerstatus
+        host.Systems.Members[0].PowerState.downcase
+      end
+
+      def poweron?
+        powerstatus == 'on'
+      end
+
+      def poweroff?
+        powerstatus == 'off'
+      end
+
+      def bootdevice
+        host.Systems.Members[0].Boot['BootSourceOverrideTarget']
+      end
+
+      def bootdevices
+        # RedFish will tell you which devices can be put in the boot source override
+        # but Foreman seems structurally arranged at present for these values to be
+        # hardcoded. I can think of a few reasons to continue with that pattern. But
+        # if in the future it becomes advantageous to do something more dynamic --
+        # this is how to find out:
+        #
+        #host.Systems.Members.Boot['BootSourceOverrideTarget@Redfish.AllowableValues']
+        #
+        # Dell gives a list like [ None, Floppy, Cd, Hdd, SDCard, Utilities,
+        #                          BiosSetup, Pxe ]
+        # HPE -        [ None, Floppy, Cd, Hdd, Usb, Utilities, BiosSetup, Pxe ]
+        # SuperMicro - [ None, Pxe, Hdd, Diags, Cd, BiosSetup, FloppyRemovableMedia,
+        #                UsbKey, UsbHdd, UsbFloppy, UsbCd, UefiCd, UefiHdd, ... ]
+        #
+        # For the four devices Foreman hardcodes, lucky for us all three use the same set
+        # of values.
+        [ 'pxe', 'disk', 'bios', 'cdrom' ]
+      end
+
+      def bootdevice=(args={ :device => nil, :reboot => false, :persistent => false })
+        devmap = { 'bios'  => 'BiosSetup',
+                   'cdrom' => 'Cd',
+                   'disk'  => 'Hdd',
+                   'pxe'   => 'Pxe' }
+
+        host.Systems.Members[0].patch(
+          payload: {
+            'Boot' => {
+              'BootSourceOverrideTarget' => devmap[args[:device]],
+              'BootSourceOverrideEnabled' => args[:persistent] ? 'Enabled' : 'Once'
+            }
+          })
+        powercycle if reboot
+      end
+
+      def bootpxe(reboot=false, persistent=false)
+        bootdevice = { :device => 'pxe', :reboot => reboot, :persistent => persistent }
+      end
+
+      def bootdisk(reboot=false, persistent=false)
+        bootdevice = { :device => 'disk', :reboot => reboot, :persistent => persistent }
+      end
+
+      def bootbios(reboot=false, persistent=false)
+        bootdevice = { :device => 'bios', :reboot => reboot, :persistent => persistent }
+      end
+
+      def bootcdrom(reboot=false, persistent=false)
+        bootdevice = { :device => 'cdrom', :reboot => reboot, :persistent => persistent }
+      end
+
+      def ip
+        host.Managers.Members[0].EthernetInterfaces.Members[0].IPv4Addresses.first['Address']
```

Comment:
This chain (and the others below) needs more testing for nil values. You can use `try` Ruby method I think since we upgraded to Ruby 2.6 already.